Source/WebCore/css/StyleProperties.cpp

```diff
@@ -984,17 +990,92 @@ String StyleProperties::getGridTemplateValue() const
     return result.toString();
 }
 
+static bool gridAutoFlowContains(const RefPtr<CSSValue>& autoFlow, CSSValueID id)
+{
+    if (is<CSSValueList>(autoFlow)) {
+        for (auto& currentValue : downcast<CSSValueList>(*autoFlow)) {
+            if (isValueID(currentValue, id))
+                return true;
+        }
+        return false;
+    }
+    return isValueID(autoFlow, id);
+}
+
 String StyleProperties::getGridValue() const
 {
     // If none of the implicit track properties have been set, then
     // this shorthand can be represented using the grid-template shorthand
     // format.
-    if (isValueID(getPropertyCSSValue(CSSPropertyGridAutoColumns), CSSValueAuto)
-        && isValueID(getPropertyCSSValue(CSSPropertyGridAutoRows), CSSValueAuto)
-        && isValueID(getPropertyCSSValue(CSSPropertyGridAutoFlow), CSSValueRow)) {
+    auto autoColumns = getPropertyCSSValue(CSSPropertyGridAutoColumns);
+    auto autoRows = getPropertyCSSValue(CSSPropertyGridAutoRows);
+    auto autoFlow = getPropertyCSSValue(CSSPropertyGridAutoFlow);
+    if (isValueID(autoColumns, CSSValueAuto)
+        && isValueID(autoRows, CSSValueAuto)
+        && isValueID(autoFlow, CSSValueRow)) {
         return getGridTemplateValue();
     }
-    return getGridShorthandValue(gridShorthand());
+
+    if (!isValueID(getPropertyCSSValue(CSSPropertyGridTemplateAreas), CSSValueNone))
+        return String();
+
+    auto rowsIndex = findPropertyIndex(CSSPropertyGridTemplateRows);
+    if (rowsIndex == -1)
+        return String();
+    auto columnsIndex = findPropertyIndex(CSSPropertyGridTemplateColumns);
+    if (columnsIndex == -1)
+        return String();
+
+    if (!autoColumns || !autoRows)
+        return String();
```

Comment:
Nit: this could be moved just after defining `autoColumns` and `autoRows`? And I guess it wouldn't hurt to check  `!autoFlow` too.